### PR TITLE
Speed up stage 1 by improving hash functions

### PR DESCRIPTION
**Pull request by SpexGuy**
A significant portion of stage 1 compilation time was spent in const_values_equal, called as part of hash map lookups.  This was called much more often than it should have been because the hash functions used for const values were not very good hashes.  This PR improves those hash functions, reducing the time to create a debug build of stage 2 from 37 seconds to 30 seconds on my computer.

I also have audited struct field accesses throughout the codebase to ensure that comptime fields are handled correctly.

### CODE REVIEW

File: src/stage1/hash_map.hpp [hunk 1]
```diff
@@ -51,7 +77,7 @@ class HashMap {
 
         if (_index_bytes == nullptr) {
             if (_entries.length < 16) {
-                _entries.append({HashFunction(key), 0, key, value});
+                _entries.append({HashFunction(MakePointer<K>::convert(key)), 0, key, value});
```
Comment (andrewrk) [inline src/stage1/hash_map.hpp:80]: forgive my lack of C++ knowledge but why not just `&key` ?

Comment (SpexGuy) [reply src/stage1/hash_map.hpp:80]: MakePointer selects an implementation based on the type.  When the key is `*T`, it's `key`.  When the key is `T`, it's `&key`.

Comment (andrewrk) [reply src/stage1/hash_map.hpp:80]: mmm I'd feel a lot more comfortable with the callsites being updated to use `&` where appropriate. One C++ feature here, one C++ feature there, and then soon enough Bjarne Soustroup is murdering your firstborn child

Comment (SpexGuy) [reply src/stage1/hash_map.hpp:80]: That doesn't work.  When the key is a pointer, we need to pass it by value.

Comment (andrewrk) [reply src/stage1/hash_map.hpp:80]: ahh I see the problem. Alright, I'm still uncomfortable with this C++ creeping in, but I think it's justified enough for me to not push back on it any more.

**Comment by andrewrk** [comment]
This branch takes me down from 20s to 17s for `./zig build -p zig-out -Dskip-install-lib-files -Dskip-non-native` with a release build of stage1.

**Review by andrewrk — APPROVED** [review]

**Pull request merged**